democracylab/oauth2.py

```diff
@@ -0,0 +1,94 @@
+"""
+Hooks for customizing login with social providers
+https://django-allauth.readthedocs.io/en/latest/advanced.html
+"""
+from allauth.account.signals import user_logged_in
+from allauth.socialaccount.adapter import DefaultSocialAccountAdapter
+from django.dispatch import receiver
+from civictechprojects.models import ProjectFile, FileCategory
+from democracylab.models import Contributor
+from django.contrib.auth.models import User
+from django.utils import timezone
+import requests
+
+
+def _get_avatar_url(sociallogin):
+    if sociallogin.account.provider.startswith('linkedin'):
+        return _get_linkedin_avatar_url(sociallogin)
+    else:
+        return sociallogin.account.get_avatar_url()
+
+def _get_linkedin_avatar_url(sociallogin):
+    response = requests.get(
+        'https://api.linkedin.com/v2/me?projection=(profilePicture(displayImage~:playableStreams))',
+        headers={'Authorization': f'Bearer {sociallogin.token}'}
+    )
+    body = response.json()
+    return body.get('profilePicture').get('displayImage~').get(
+        'elements')[-1].get('identifiers')[0].get('identifier')
+
+def _update_user_avatar_url(sociallogin):
+    owner = sociallogin.user.contributor
+    user_avatar_url = _get_avatar_url(sociallogin)
+    if user_avatar_url:
+        file_json = {
+            'publicUrl': user_avatar_url,
+            'file_user': owner,
+            'file_category': FileCategory.THUMBNAIL.value,
+            'visibility': 'PUBLIC',
+            'fileName': f'{owner.first_name}{owner.last_name}_thumbnail.{sociallogin.account.provider} avatar',
+            'key': f'{sociallogin.account.provider}/{owner.username}'
+        }
+        ProjectFile.replace_single_file(owner, FileCategory.THUMBNAIL, file_json)
+
+
+class SocialAccountAdapter(DefaultSocialAccountAdapter):
+    def new_user(self, request, sociallogin):
+        email = sociallogin.account.get_provider().extract_common_fields(
+                                                   sociallogin.account.extra_data).get('email').lower()
+        try:
+            # This may NOT necessarily be a new user
+            user = User.objects.get(username=email)
+            # Preserve current password (sociallogin assigns an unusable password)
+            if user.has_usable_password():
+                sociallogin.account.extra_data.update(password=user.password)
+            return Contributor.objects.get_by_natural_key(user.username)
+
+        except User.DoesNotExist:
```

Comment:
In general, exceptions should be reserved for 'exceptional' circumstances.  In this case, we expect people without prior DemocracyLab accounts to sign in via social media, so we should try to avoid using exceptions for it.

Instead, we can do something like
user_match=User.objects.filter(username=email)
if user_match.exists():
	# code for existing user
else:
	# code for new user

Reply:
Yes that's true in general, but going to the database to see if something exists, and then going back to the database to actually retrieve it feels much less elegant than handling the exception we expect. The exception can be viewed as just another return value.

I think the compromise would be `user = User.objects.filter(username=email).first()`